src/foam/nanos/ruler/RetryManager.java

```diff
@@ -39,6 +46,7 @@ public void run() {
               latch_.countDown();
             }
           } catch (Exception ex) {
+            exception_ = ex;
```

Comment:
you should only keep the first exception .. so how about: 
```suggestion
            exception_ = (exception_ == null) ? ex : exception_;
```

Reply:
yes, agree.

Reply:
The first exception is already log in the RuleEngine class. This is the last exception when we stop retrying.

Reply:
Then I'm ok with it.

Reply:
I'm ok with it too.